src/lib/dnssd/IncrementalResolve.cpp

```diff
@@ -0,0 +1,383 @@
+constexpr QNamePart kCommissionerSuffix[]          = { kCommissionerServiceName, kCommissionProtocol, kLocalDomain };
+constexpr QNamePart kCommissionableSubTypeSuffix[] = { kSubtypeServiceNamePart, kCommissionableServiceName, kCommissionProtocol,
+                                                       kLocalDomain };
+constexpr QNamePart kCommissionerSubTypeSuffix[]   = { kSubtypeServiceNamePart, kCommissionerServiceName, kCommissionProtocol,
+                                                     kLocalDomain };
+
+ServiceNameType ComputeServiceNameType(SerializedQNameIterator name)
+{
+    // SRV record names look like:
+    //   <fabric>-<node>._matter._tcp.local  (operational)
+    //   <instance>._matterc._udp.local  (commissionable)
+    //   <instance>._matterd._udp.local  (commissioner)
+
+    // Get fabric/node or instance name first
+    // This skips over that value
+    if (!name.Next() || !name.IsValid())
+    {
+        // missing required components - empty service name
+        return ServiceNameType::kInvalid;
+    }
+
+    if (name == kOperationalSuffix)
+    {
+        return ServiceNameType::kOperational;
+    }
+    else if (name == kCommissionableSuffix)
+    {
+        return ServiceNameType::kCommissionable;
+    }
+    else if (name == kCommissionerSuffix)
+    {
+        return ServiceNameType::kCommissioner;
+    }
+    else
+    {
+        return ServiceNameType::kInvalid;
+    }
+}
+
+/// Checks if the name is of the form <something>._sub._matter(c|d)._udp.local
+bool IsCommissionSubtype(SerializedQNameIterator name)
+{
+    if (!name.Next() || !name.IsValid())
+    {
+        // subtype should be a prefix
+        return false;
+    }
+
+    return (name == kCommissionerSubTypeSuffix) || (name == kCommissionableSubTypeSuffix);
+}
+
+} // namespace
+
+CHIP_ERROR StoredServerName::Set(SerializedQNameIterator value)
+{
+    chip::Encoding::BigEndian::BufferWriter output(mBuffer, sizeof(mBuffer));
+    RecordWriter writer(&output);
+
+    writer.WriteQName(value);
+
+    if (!writer.Fit())
+    {
+        Clear();
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+SerializedQNameIterator StoredServerName::Get() const
+{
+    return SerializedQNameIterator(BytesRange(mBuffer, mBuffer + sizeof(mBuffer)), mBuffer);
+}
+
+CHIP_ERROR IncrementalResolver::InitializeParsing(mdns::Minimal::SerializedQNameIterator name, const mdns::Minimal::SrvRecord & srv)
+{
+    mSpecificResolutionData = SpecificParseData();
+    mCommonResolutionData.Reset();
+
+    ReturnErrorOnFailure(mRecordName.Set(name));
+    ReturnErrorOnFailure(mServerName.Set(srv.GetName()));
+    mCommonResolutionData.port = srv.GetPort();
+
+    {
+        // TODO: Chip code historically seems to assume that the host name is of the
+        // form "<MAC or 802.15.4 Extended Address in hex>.local" and only saves the first part.
+        //
+        // This should not be needed as server name should not be relevant once parsed.
```

Comment:
The hostname appears in A/AAAA records, so the comment is confusing re: "should not be needed". Also curious why you call it "serverName". "SRV == Service" not "Server"